daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -2143,6 +2143,31 @@ private[lf] object SBuiltin {
     }
   }
 
+  /** $dynamicExercise[T, C, R] :: HasDynamicExercise T C R => ContractId T ->  C -> Update R */
+  final case class SBUDynamicExercise(
+      templateId: TypeConName,
+      choice: ChoiceName,
+  ) extends UpdateBuiltin(2) {
+    override protected def executeUpdate(
+        args: util.ArrayList[SValue],
+        machine: UpdateMachine,
+    ): Control[Question.Update] = {
+      Control.Question(
+        Question.Update.NeedPackageId(
+          module = templateId.qualifiedName.module,
+          pid0 = templateId.packageId,
+          callback = pid => {
+            val e = SEApp(
+              SEVal(TemplateChoiceDefRef(templateId.copy(packageId = pid), choice)),
+              args.toArray(Array.empty[SValue]),
+            )
```

Comment:
I think it should work with only 
```suggestion
            val e = SEVal(TemplateChoiceDefRef(templateId.copy(packageId = pid), choice))
```
In fact I think is a bad idea to apply something to no arguments.

Reply:
ah, the `empty` part is confusing; it's needed to get the right conversion so `TemplateChoiceDefRef` is applied to `args`, _not_ the empty array

Reply:
if you know of a better way to convert from `java.util.ArrayList[x]` to `Array[x]` I'd be happy to remove `Array.empty`

Reply:
You are right.

Reply:
Maybe `args.asScala.toArray` is more readable.

Reply:
We could also compile make the builtin with only one argument (the token) and compile the 

`UpdateDynamicExercise(..) cid token` as `SBUDynamicExercise(..) token cid token` with arity of `SBUDynamicExercise` being 1. 

What is you opinion @nickchapman-da ?